packages/headless/src/features/search-parameters/search-parameter-serializer.test.ts

```diff
@@ -5,15 +5,34 @@ describe('buildSearchParameterSerializer', () => {
   describe('#serialize', () => {
     const {serialize} = buildSearchParameterSerializer();
 
-    it('encodes a record with a single key and string value', () => {
+    it('serializes a record with a single key and string value', () => {
       const result = serialize({q: 'a'});
       expect(result).toBe('q=a');
     });
 
-    it('encodes a record with a multiple keys and string values', () => {
+    it('serializes a record with a multiple keys and string values', () => {
       const result = serialize({q: 'a', enableQuerySyntax: true});
       expect(result).toBe('q=a&enableQuerySyntax=true');
     });
+
+    it('does not include invalid keys', () => {
+      const invalidKeys = {invalid: 'b'};
+      const result = serialize({q: 'a', ...invalidKeys});
+
+      expect(result).toBe('q=a');
+    });
+
+    it('serializes the #f parameter correctly', () => {
```

Comment:
Missing quite a bit of tests for any sensitive character that would trip up:  

* Any Facet values that contains any URL sensitive char (eg: values should be ULR encoded). This include comma, bracket, `=` sign, etc.. Any character that could trip up the decoding part.
 
* Facet ID is also not URL encoded, and this is essentially user input. Bunch of characters in Facet ID would break the decoding part.

This URL encode comment is valid for all parameters value, by the way.
And for any parameter key that we do not control 100%.

Reply:
I think the better approach is not to encode/decode in the serializer because a) knowing what characters to encode, and b) knowing how they should be encoded, depends on the environment where headless is being used.

I also think headless should not bundle an encoder/decoder dependency with it's serializer to keep the payload light.

One argument might be that initializing the serializer should force an encoder/decoder pair to remind developers about this. I did not find this friendly when implementing [the example inside Atomic](https://github.com/coveo/ui-kit/blob/master/packages/atomic/src/components/atomic-search-interface/atomic-search-interface.tsx#L107).

When setting the url hash, the browser auto-encodes for you, so the encoder is not necessary. We could make the decoder a required option but it feels less elegant/not symmetrical to have a required decoder option.

Reply:
So... if I have a facet value which contains a comma, what do I do as a user of headless ? That's not clear to me 🤔

Reply:
Just to close the loop, yes it's true, if a value contains a delimiter, then the deserializer will produce the wrong output. There is [KIT-296](https://coveord.atlassian.net/browse/KIT-296) to explore these edge-cases.